---
title: "Module11_Homework"
author: "Nick Wlodychak"
date: "2023-12-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

Cancer genomics is a rapidly growing and expanding field of bioinformatics investigating genetic abnormalities in cancer cells. Data architecture and formatting is critical in a disease with so many intricacies and interactions. The Mutation Annotation Format (MAF) is a commonly used format that stores aggregated mutation information from Variant Call Format (VCF) files.

MAF files are generated on a project-level and contain essential information such as variant classification and amino acid changes. MAF are used broadly applicable in cancer genomics and more generally used to study somatic variants in large datasets, such as those from The Cancer Genome Atlas (TCGA). MAF files are used to store somatic variants detected in over 30 different cancers with sample sizes of each cancer type being over 200, making them critical for comprehensive analysis and visualization of somatic variants in cancer

In this tutorial we use `maftools`, a powerful R library utilized for the summation of MAF files, visualization of somatic mutations through oncoplots (waterfall plots), and analysis of oncogenomic data-sets. `maftools` possesses categorical visualization and analysis modules.

# **Reading and summarizing `maf` files**

The first step in our tutorial is to read our `maf` file with the function **`read.maf`**. This function summarizes `maf` in various ways, and stores the summarized data as an MAF object. Additionally the function can integrate copy number data from GISTIC or a custom table containing gene name, sample name, and copy number status.

Following incorporation of our `maf` file as a variable, we can summarize and view the data set with functions such as `getSampleSummary` and `getFields`. These functions can help with downstream variable assignment and data integrity.

```{r 6 Reading and summarizing maf files}
library(maftools)
#path to TCGA LAML MAF file
laml.maf = system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools')
#clinical information containing survival information and histology. This is optional
laml.clin = system.file('extdata', 'tcga_laml_annot.tsv', package = 'maftools')
laml = read.maf(maf = laml.maf, clinicalData = laml.clin)
#Typing laml shows basic summary of MAF file.
```

```{r 6.3 MAF Object}
laml
#Shows sample summry.
getSampleSummary(laml)
#Shows gene summary.
getGeneSummary(laml)
#shows clinical data associated with samples
getClinicalData(laml)
#Shows all fields in MAF
getFields(laml)
#Writes maf summary to an output file with basename laml.
write.mafSummary(maf = laml, basename = 'laml')
```

# **Visualization modules**

Good data visualization is critical to the success of any bioinformatician. Conveying significance and values of observations assits in general understanding of a dataset. the maftools package posesses a robust suite of visualization tools that plot a variety of variables including `VAF` and mutational loads

Below we show a group of plots generated via `plotmafSummary`. Each plot is a different method of representing the data to give a comprehensive view of the mutation landscape within the studied samples. This is useful understanding patterns and frequencies of various types of genomic alterations in cancer. Higlights include:

-   **Variant Classification** for determination of mutation type.

-   **SNV Class** plotting of single nucleotide variants (SNVs) showing the different base changes

-   **Top 10 Mutated Genes** lists the top 10 genes with mutations.

```{r 7.1 Plotting MAF summary.}
plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
```

### Oncoplots

Below we show an oncoplot which is similar to a histogram for 193 samples in the data-set. Mutations are altered in 141 out of 193 samples, accounting for 73.06% of the cohort. Different mutations are represented in the color key which the top histogram shows the Tumor mutation burden (TMB) across the samples. Genes are stacked as the top mutated genes found in these samples.

This type of plot is useful for observing mutational load within a population and also analyzing the frequency of gene specific mutations and their genomic loci.

```{r 7.2 Oncoplots}
#oncoplot for top ten mutated genes.
oncoplot(maf = laml, top = 10)
```

### Transition and Transversions

This boxplots for displays the common nucleotide changes (C\>T, C\>A, T\>A) and the transition/transversion (Ti/Tv) ratio in mutations. In our plot, a high number of C\>T transitions suggest a possible signature of UV-light induced damage or spontaneous deamination of 5-methylcytosine.

```{r 7.3 Transition and Transversions.}
laml.titv = titv(maf = laml, plot = FALSE, useSyn = TRUE)
#plot titv summary
plotTiTv(res = laml.titv)
```

### Rainfall plots

This graph illustrates the distribution of mutations across all chromosomes in a single sample. Each dot in a rainfall plot represents a mutation, with its color indicating the type of nucleotide change. The y-axis shows the log10 of the inter-event distance --- tight clusters may suggest kataegis or localized hypermutation.

```{r 7.5 Rainfall plots}
brca <- system.file("extdata", "brca.maf.gz", package = "maftools")
brca = read.maf(maf = brca, verbose = FALSE)
rainfallPlot(maf = brca, detectChangePoints = TRUE, pointSize = 0.4)
```

### Mutation load

This plot displays the tumor mutation burden (TMB) across various cancer types, with each dot representing a sample. The y-axis is on a logarithmic scale, indicating the number of mutations per megabase (MB) of tumor DNA. The red line indicates the median TMB for each cancer type, displaying the variability in mutation load across different cancers subsets.

```{r 7.6 Compare mutation load against TCGA cohorts}
laml.mutload = tcgaCompare(maf = laml, cohortName = 'Example-LAML', logscale = TRUE, capture_size = 50)
```

### Plotting VAF

```{r 7.7 Plotting VAF}
plotVaf(maf = laml, vafCol = 'i_TumorVAF_WU')
```

# Analysis Modules

### Somatic Interactions

The plot below plot displays the statistical relationship between mutations in different genes. The shading of each cell indicates the significance value with the strength of evidence for co-occurrence or mutual exclusivity of mutations.

```{r 9 Analysis}
somaticInteractions(maf = laml, top = 25, pvalue = c(0.05, 0.1))
```

### Oncodrive

The function `oncodrive` identifies cancer genes (driver) from a given MAF. This function is based on cancer causing variants being enriched at some specific loci. By analyzing such positions we can identify cancer causing genes.

This plotting then plots the results as scatter plot size being indicative of the number of hot-spots in the gene clusters.

```{r 9.2 Detecting cancer driver genes based on positional clustering}
laml.sig = oncodrive(maf = laml, AACol = 'Protein_Change', minMut = 5, pvalMethod = 'zscore')
head(laml.sig)
plotOncodrive(res = laml.sig, fdrCutOff = 0.1, useFraction = TRUE, labelSize = 0.5)
```

### **Adding and summarizing pfam domains**

The `pfamDomain` function identifies amino acid changes according to the domains that are affected. From these mutations we can identify which protein domain is most affected within a cancer cohort.

```{r 9.3 Adding and summarizing pfam domains}
laml.pfam = pfamDomains(maf = laml, AACol = 'Protein_Change', top = 10)
laml.pfam$proteinSummary[,1:7, with = FALSE]
laml.pfam$domainSummary[,1:3, with = FALSE]
```

### Survival Analysis

The survival curve compares the survival probability over time between two groups with DMT3A (a methyltransferase) - mutant genes (Mutant) and wild-type (WT) genes. The significantly different survival curves, along with the P-value (7e-04) and hazard ratio (HR: 1.98), indicate that having the mutant gene is associated with a worse survival outcome compared to the wild-type, suggesting the mutation's potential role as a prognostic indicator.

```{r 9.4 Survival analysis}
mafSurvival(maf = laml, genes = 'DNMT3A', time = 'days_to_last_followup', Status = 'Overall_Survival_Status', isTCGA = TRUE)
prog_geneset = survGroup(maf = laml, top = 20, geneSetSize = 2, time = "days_to_last_followup", Status = "Overall_Survival_Status", verbose = FALSE)
print(prog_geneset)
mafSurvGroup(maf = laml, geneSet = c("DNMT3A", "FLT3"), time = "days_to_last_followup", Status = "Overall_Survival_Status")
```

### Relapse versus Primary Comparison Plots

This plot compares the frequency of certain gene mutations in relapsed versus primary cancer cases, represented by the odds ratio (OR). Genes like ***PML***, ***RARA***, and ***RUNX1*** show a higher mutation frequency in relapse (indicated by OR less than 1), suggesting these mutations could be associated with relapse. ***FLT3*** shows an OR greater than 1, indicating more frequent mutations in primary cases.

```{r 9.5 Comparing two cohorts (MAFs)}
#Primary APL MAF
primary.apl = system.file("extdata", "APL_primary.maf.gz", package = "maftools")
primary.apl = read.maf(maf = primary.apl)
#Relapse APL MAF
relapse.apl = system.file("extdata", "APL_relapse.maf.gz", package = "maftools")
relapse.apl = read.maf(maf = relapse.apl)
#Considering only genes which are mutated in at-least in 5 samples in one of the cohort to avoid bias due to genes mutated in single sample.
pt.vs.rt <- mafCompare(m1 = primary.apl, m2 = relapse.apl, m1Name = 'Primary', m2Name = 'Relapse', minMut = 5)
print(pt.vs.rt)
```

### Forest Plot and Co-onco plots

The bar plots display the proportion of different mutation types in primary and relapse Acute Promyelocytic Leukemia (APL).

Mutations such as missense and nonsense mutations are represented, with a notable difference in the distribution between primary and relapse cases, which could imply a change in the mutational landscape post-treatment.

```{r 9.5.1 Forest plots}
forestPlot(mafCompareRes = pt.vs.rt, pVal = 0.1)
```

```{r 9.5.2 Co-onco plots}
genes = c("PML", "RARA", "RUNX1", "ARID1B", "FLT3")
coOncoplot(m1 = primary.apl, m2 = relapse.apl, m1Name = 'PrimaryAPL', m2Name = 'RelapseAPL', genes = genes, removeNonMutated = TRUE)
```

### Lollipop plot

This lollipop plot shows the distribution and types of mutations in the PML gene for primary and relapse cases. A higher mutation frequency in relapse cases (18.97%) versus primary cases (0.81%) suggests these mutations could play a role in disease progression or treatment resistance.

```{r 9.5.4 Lollipop plot-2}
lollipopPlot2(m1 = primary.apl, m2 = relapse.apl, gene = "PML", AACol1 = "amino_acid_change", AACol2 = "amino_acid_change", m1_name = "Primary", m2_name = "Relapse")
```

### Clinical enrichment analysis

This plot illustrates the variant allele frequencies of specific gene mutations across different French-American-British (FAB) classifications of leukemia. The distribution of mutations among various classifications could suggest differing roles of these genes in the subtypes of leukemia.

```{r 9.6 Clinical enrichment analysis}
fab.ce = clinicalEnrichment(maf = laml, clinicalFeature = 'FAB_classification')
plotEnrichmentResults(enrich_res = fab.ce, pVal = 0.05, geneFontSize = 0.5, annoFontSize = 0.6)
```

### Drug-Gene Interactions

The word cloud categorizes genes by their potential as drug targets, with the size indicating the number of genes within each category. Genes involved in pathways like transcription factor binding, tyrosine kinase, and DNA repair are highlighted, suggesting these are key areas for targeted therapies.

```{r 9.7 Drug-Gene Interactions}
dgi = drugInteractions(maf = laml, fontSize = 0.75)
dnmt3a.dgi = drugInteractions(genes = "DNMT3A", drugs = TRUE)
dnmt3a.dgi[,.(Gene, interaction_types, drug_name, drug_claim_name)]
```

### Oncogenic Signaling Pathways

This plot shows the fraction of different cancer-related pathways affected in a sample set, comparing the fraction within a specific subset (e.g., RTK-RAS pathway) to the overall sample population. This plot can assist with identifying those pathways most significantly affected by cancer diagnosis.

```{r 9.8 Oncogenic Signaling Pathways}
OncogenicPathways(maf = laml)
PlotOncogenicPathways(maf = laml, pathways = "RTK-RAS")
```

# Conclusion

Through the provided `maftools` tutorial, we've gained the toolset to manipulate and interpret cancer data. We've learned how to read and summarize Mutation Annotation Format (`MAF`) files, which are critical for storing extensive genomic mutation data. The various plots and visualizations, such as oncoplots and waterfall plots are important in the exploration of mutational patterns and mutated genes identification within a cohort.

Analysis tools within `maftools`, including survival analysis and tumor mutation burden calculation, allow us to better understand the significance of cancer causing genes within a diagnosis.

By utilizing these tools, we can extract meaningful conclusions from complex genomic or multiomic data-sets. These insights are indespensable for improving surviveability and prescribing personalized medicine in oncology. For example, identifying recurrent mutations in a patient's tumor profile could guide the selection of targeted therapies that are more likely to be effective against their specific cancer type. Survival analysis and mutation burden estimation are critical for prognosis and making informed decisions about dose and drug intensity.

While `maftools` provides a solid foundation for cancer genomic analysis, a more robust cancer identification pipeline might integrate additional data-sets (transcriptomic or methylomic), and utilize machine learning algorithms to predict treatment outcomes or identify novel therapeutic targets. Complementary tools like Genome Analysis Toolkit (GATK), known for its sophisticated variant discovery in high-throughput sequencing data, can complement the capabilities of `maftools`. These tool-sets combined can increase the power of patient care and treatment.

# References

1.  Cancer Genome Atlas Research, N. Genomic and epigenomic landscapes of adult de novo acute myeloid leukemia. N Engl J Med 368, 2059-74 (2013).

2.  David Tamborero, Abel Gonzalez-Perez, Nuria Lopez-Bigas, OncodriveCLUST: exploiting the positional clustering of somatic mutations to identify cancer genes, *Bioinformatics*, Volume 29, Issue 18, September 2013, Pages 2238--2244, <https://doi.org/10.1093/bioinformatics/btt395>

3.  Mayakonda A, Lin DC, Assenov Y, Plass C, Koeffler HP. 2018. Maftools: efficient and comprehensive analysis of somatic variants in cancer. [Genome Resarch. PMID: 30341162](https://doi.org/10.1101/gr.239244.118)
